src/http.c

```diff
@@ -150,75 +155,87 @@ int http_receive(
-		if (n > 0) {
-			log_debug_details("%s:\n%s\n", __func__, buffer);
-			const char *eoh;
-
-			bytes_read += n;
-
-			if (!header_size) {
-				/* Did we see the header end? Then get the body size. */
-				eoh = memmem(buffer, bytes_read, "\r\n\r\n", 4);
-				if (eoh) {
-					const char *header;
-
-					header = find_header(
-					                 buffer,
-					                 "Content-Length: ", BUFSZ);
-					header_size = eoh - buffer + 4;
-					if (header)
-						content_size = atoi(header);
-
-					if (find_header(buffer,
-					                "Transfer-Encoding: chunked",
-					                BUFSZ))
-						chunked = 1;
-				}
+	while (1) {
+		int n;
+
+		while ((n = safe_ssl_read(tunnel->ssl_handle,
+		                          (uint8_t *) buffer + bytes_read,
+		                          capacity - bytes_read)) == ERR_SSL_AGAIN)
+			;
+		if (n < 0) {
+			log_error("Error reading from SSL connection (%s).\n",
+			          err_ssl_str(n));
+			free(buffer);
+			return ERR_HTTP_SSL;
+		}
+		bytes_read += n;
+
+		log_debug_details("%s:\n%s\n", __func__, buffer);
+
+		if (!header_size) {
+			/* Have we reached the end of the HTTP header? */
+			static const char EOH[4] = "\r\n\r\n";
+			const char *eoh = memmem(buffer, bytes_read,
+			                         EOH, sizeof(EOH));
+
+			if (eoh) {
+				header_size = eoh - buffer + sizeof(EOH);
+
+				/* Get the body size. */
+				const char *header = find_header(buffer,
+				                                 "Content-Length: ",
+				                                 header_size);
+
+				if (header)
+					content_size = atoi(header);
+
+				if (find_header(buffer,
+				                "Transfer-Encoding: chunked",
+				                header_size))
+					chunked = 1;
 			}
+		}
 
-			if (header_size) {
-				/* We saw the whole header, is the body done as well? */
-				if (chunked) {
-					/* Last chunk terminator. Done naively. */
-					if (bytes_read >= 7 &&
-					    !memcmp(&buffer[bytes_read - 7],
-					            "\r\n0\r\n\r\n", 7))
-						break;
-				} else {
-					if (bytes_read >= header_size + content_size)
-						break;
-				}
+		if (header_size) {
+			/* Have we reached the end of the HTTP body? */
+			if (chunked) {
+				static const char EOB[7] = "\r\n0\r\n\r\n";
+
+				/* Last chunk terminator. Done naively. */
+				if (bytes_read >= sizeof(EOB) &&
+				    !memcmp(&buffer[bytes_read - sizeof(EOB)],
+				            EOB, sizeof(EOB)))
+					break;
+			} else {
+				if (bytes_read >= header_size + content_size)
+					break;
 			}
+		}
 
-			if (bytes_read == BUFSZ - 1) {
-				log_warn("Response too big\n");
+		/* expand the buffer if necessary */
+		if (bytes_read == capacity) {
+			char *new_buffer;
+
+			assert(UINT32_MAX / capacity >= 2);
+			capacity *= 2;
+			new_buffer = realloc(buffer, capacity);
+			if (new_buffer == NULL) {
 				free(buffer);
-				return ERR_HTTP_SSL;
+				return ERR_HTTP_NO_MEM;
 			}
+			buffer = new_buffer;
 		}
-	} while (n >= 0);
-
```

Comment:
the fact that you drop this condition is because you handle ssl errors in the code above, right?

Reply:
Absolutely, I have added a `return` inside the loop, triggered by `if (n < 0)`:
https://github.com/DimitriPapadopoulos/openfortivpn/blob/c2c467ed067372a0568ee7e7f1bf409034e2467f/src/http.c#L176
This is equivalent to continuing `while (n >= 0)` but in my view has the advantage of showing explicitly which part of the code is responsible for errors.

Note that I have added a test for `n == ERR_SSL_AGAIN` (⇔ `n == 0`) early in the loop:
https://github.com/DimitriPapadopoulos/openfortivpn/blob/c2c467ed067372a0568ee7e7f1bf409034e2467f/src/http.c#L172-L175
I believe it improves readability through proximity of conditional expressions and assignment of the variables `n` these conditionals operate on. Additionally this helps remove the `if (n> 0)` conditional in the loop since we now handle the `n < 0` and `n == 0` cases earlier in the loop.

Reply:
The main change is that we now **return** from the function if `n < 0`.

We would previously only **leave the loop** if `n < 0`. Then immediately after the loop we would return only if `!header` - discarding possible SSL violations.

Reply:
thank you for the explanations